compiler/dyno/test/resolution/testReturnTypes.cpp

```diff
@@ -0,0 +1,345 @@
+/*
+ * Copyright 2021-2022 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "chpl/parsing/parsing-queries.h"
+#include "chpl/resolution/resolution-queries.h"
+#include "chpl/resolution/scope-queries.h"
+#include "chpl/types/all-types.h"
+#include "chpl/uast/Identifier.h"
+#include "chpl/uast/Module.h"
+#include "chpl/uast/Record.h"
+#include "chpl/uast/Variable.h"
+#include "common.h"
+
+#include <cassert>
+
+struct ReturnVariant {
+  bool literal;
+  std::string intent;
+  std::string type;
+  std::string value;
+  bool generic;
+
+  ReturnVariant(std::string intent, std::string type, bool generic=false)
+    : literal(false), intent(std::move(intent)), type(std::move(type)),
+      value(""), generic(generic) {}
+
+  ReturnVariant(std::string value) : literal(true), intent(""), type(""), value(std::move(value)) {}
+};
+
+std::string buildProgram(std::string intent, std::string selectorValue, const std::vector<ReturnVariant>& variants) {
```

Comment:
IMO this test would be clearer if it called commonType directly. I'm finding it really hard to understand what is going on here. Now, you might say, but  it's annoying to construct all of the types to test! You could parse a program with variables declared with the different types & then call commonType on the resulting types.